src/kOS.Safe/Encapsulation/Structure.cs

```diff
@@ -337,20 +337,10 @@ public static Structure FromPrimitiveWithAssert(object value)
 
         public static object ToPrimitive(object value)
         {
-            var scalarValue = value as ScalarValue;
-            if (scalarValue != null)
+            var primitive = value as PrimitiveStructure;
+            if (primitive != null)
             {
-                return scalarValue.Value;
-            }
-            var booleanValue = value as BooleanValue;
-            if (booleanValue != null)
-            {
-                return booleanValue.Value;
-            }
-            var stringValue = value as StringValue;
-            if (stringValue != null)
-            {
-                return stringValue.ToString();
+                return primitive.ToPrimitive();
```

Comment:
IMO my change simplifies the fact that a given type is a primitive. Previously that information was stored in 2 places - `ISerializableValue` - used by serialization and `Structure.ToPrimitive`. This could cause bugs where a type was handled by `ToPrimitive`, but not marked with `ISerializableValue` or vice-versa (as actually is the case now with `BooleanValue`). Now if a type is a `PrimitiveStructure` it must implement `ToPrimitive`.